src/modules/rlm_sql/rlm_sql.c

```diff
@@ -779,6 +779,86 @@ static int sql_get_grouplist(rlm_sql_t const *inst, rlm_sql_handle_t **handle, R
 	return num_groups;
 }
 
+static rlm_rcode_t sql_get_grouplist_resume(REQUEST *request, void *instance, void *thread, void *ctx);
+
+/*
+ * Get group list
+ * Yield if SQL driver supports async queries.
+ */
+static rlm_rcode_t sql_get_grouplist_async(REQUEST *request, void *instance, void *thread, void *ctx)
+{
+	rlm_sql_thread_group_t *group_ctx = ctx;
+	rlm_sql_t const *inst = instance;
+
+	int rcode;
+
+	/* NOTE: sql_set_user should have been run before calling this function */
+
+	if (!inst->config->groupmemb_query || !*inst->config->groupmemb_query) {
+		group_ctx->sql_getvpdata_ctx.rows = 0;
+		return RLM_MODULE_OK;
+	}
+	if (xlat_aeval(request, &group_ctx->sql_getvpdata_ctx.query, request, inst->config->groupmemb_query,
+			 inst->sql_escape_func, *group_ctx->sql_getvpdata_ctx.handle) < 0) {
+		group_ctx->sql_getvpdata_ctx.rows = -1;
+		return RLM_MODULE_OK;
+	}
+
+	rcode = rlm_sql_select_query(inst, request, group_ctx->sql_getvpdata_ctx.handle, group_ctx->sql_getvpdata_ctx.query);
+
+	switch(rcode) {
```

Comment:
I guess this is OK, but for general sanity and consistency, it might be better to return control back to unlang even if the result was RLM_SQL_OK, and let it call the next state in the state machine.  Might need some tweaking to add some sort of NULL I/O handler.

I guess I don't have solid reasons for this... just:
- Seems odd to me to do a state transition by adding frames onto the C stack.
- We're very likely going to want to add stats that go down to a per-state resolution.  Having states visible for some SQL drivers and not for others is going to confuse someone.